dotnet/src/Connectors/Connectors.Memory.Kusto/KustoMemoryStore.cs

```diff
@@ -0,0 +1,393 @@
+                this._database,
+                similiaryQuery,
+                GetClientRequestProperties(),
+                cancellationToken
+            ).ConfigureAwait(false);
+        while (reader.Read())
+        {
+            var key = reader.GetString(0);
+            var metadata = reader.GetString(1);
+            DateTime? timestamp = reader.IsDBNull(2) ? null : reader.GetDateTime(2);
+            var similarity = reader.GetDouble(3);
+            var _embedding = withEmbeddings ? reader.GetString(4) : default;
+
+            var kustoRecord = new KustoMemoryRecord(key, metadata, _embedding, timestamp);
+
+            yield return (kustoRecord.ToMemoryRecord(), similarity);
+        }
+    }
+
+    /// <inheritdoc/>
+    public Task RemoveAsync(string collectionName, string key, CancellationToken cancellationToken = default) => this.RemoveBatchAsync(collectionName, new[] { key }, cancellationToken);
+
+    /// <inheritdoc/>
+    public async Task RemoveBatchAsync(string collectionName, IEnumerable<string> keys, CancellationToken cancellationToken = default)
+    {
+        if (keys != null)
+        {
+            var keysString = string.Join(',', keys.Select(k => $"'{k}'"));
+            using var resp = await this._adminClient
+                .ExecuteControlCommandAsync(
+                    this._database,
+                    CslCommandGenerator.GenerateDeleteTableRecordsCommand(GetTableName(collectionName), $"{GetTableName(collectionName)} | where Key in ({keysString})"),
+                    GetClientRequestProperties()
+                ).ConfigureAwait(false);
+        }
+    }
+
+    /// <inheritdoc/>
+    public async Task<string> UpsertAsync(string collectionName, MemoryRecord record, CancellationToken cancellationToken = default)
+    {
+        var result = this.UpsertBatchAsync(collectionName, new[] { record }, cancellationToken);
+        return await result.FirstOrDefaultAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<string> UpsertBatchAsync(string collectionName, IEnumerable<MemoryRecord> records, CancellationToken cancellationToken = default)
+    {
+        // Upserts don't exist in Kusto, as it is an append only store.
+        // However, since we have an easy PK, we can just insert a new record, as our query always picks the latest row of that PK.
+        // An intresting scenario is deletion after a lot of "Upserts". This could be a heavy operation, as in theory there could be many old version that we now need to remove.
+        // Plus, deleting those records is a "soft delete" operation.
+        // For simplicty purposes (and under the assumption that in most systems, upserts are rare), we will ignore the potential build-up of "garbage" records,
+        // as Kusto is generally good with large amounts of data.
+
+        var inlineIngestionCommand = $".ingest inline into table {GetTableName(collectionName)} <|\n";
```

Comment:
>inlineIngestionCommand [](http://example.com/codeflow?start=12&length=22)

Doesn't CslCommandGenerator already have this template?

Reply:
yes. fixed. thank you!